[#19] Support REST Api for sending configs and commands to devices (Device Communication API)

gdimitropoulos-sotec opened the pull request:
Issue #19  is part of qualifying hono as a replacement for Google IoT Core customers (IoT Core is being retired in August 2023). 
See also: https://github.com/eclipse-hono/hono/issues/3441 & https://github.com/eclipse-hono/hono/issues/3442

At this PR we have created the Device Communication API and the needed database table. At the moment the API is not able to communicate with the Hono command router and that will be achieved when the pub-sub communication is implemented in future PR.

Signed-off-by: g.dimitropoulos[g.dimitropoulos@sotec.eu](mailto:g.dimitropoulos@sotec.eu)

gdimitropoulos-sotec committed:
create device communication api

Signed-off-by: g.dimitropoulos <g.dimitropoulos@sotec.eu>

gdimitropoulos-sotec committed:
code refactoring and  update readme

gdimitropoulos-sotec committed:
update README.md and add database functionality

gdimitropoulos-sotec committed:
remove entity package

gdimitropoulos-sotec committed:
change database naming

calohmn commented:
Thanks for the PR.
As a general comment, before diving into the specifics of this component, it would be good if the added module here would adhere to the code style used in Hono. To add corresponding checks, you could add the corresponding maven plugin configurations 
https://github.com/eclipse-hono/hono-extras/blob/2be112bd43f4a31c75354666a194170ddd2a2dc3/protocol-gateway/azure-mqtt-protocol-gateway/pom.xml#L212-L242

to the pom of the module. To do the corresponding local IDE setup, there is a [chapter](https://www.eclipse.org/hono/docs/dev-guide/building_hono/#ide-setup) in the Hono documentation.

calohmn commented on device-communication/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,36 @@
+  container-image:
+    builder: docker
+    build: true
+    push: true
```
I think `push` should by default be `false` here, so that `mvn clean install` doesn't try to push the docker image. 
In order to enable pushing the image for a `mvn` invocation, I could imagine introducing a corresponding Maven profile, by which the above value gets set to `true`.

gdimitropoulos-sotec replied:
Done

calohmn commented on device-communication/src/main/java/org/eclipse/hono/communication/core/utils/DbUtils.java:
```diff
@@ -0,0 +1,67 @@
+        pool.getConnection(connection -> {
+            if (connection.failed()) {
+                log.error(String.format(connectionFailedMsg, connection.cause().getMessage()));
+                System.exit(-1);
```
When first testing the application and starting up without a corresponding database, this `System.exit` caused a deadlock (`'Quarkus#blockingExit' was called on the main thread. This will result in deadlocking the application! [Error Occurred After Shutdown]`). Using `Quarkus.asyncExit(-1)` instead would prevent that.

As a general note:
In the Hono components, the connection establishment to DB / messaging network components is decoupled from the (Quarkus) application startup. There are also retries being done. When a connection hasn't been established yet, the corresponding readiness check fails, in the Kubernetes context meaning the component isn't considered for incoming requests yet and is eventually getting killed if it keeps being not ready.
For the `createDbClient` method here, this would mean returning a `Future<PyPool>` instead. I'm still having a look at the overall structure here to get an idea of what else would need to be changed.

gdimitropoulos-sotec replied:
I have replaced the system.exit with  Quarkus.asyncExit.

gdimitropoulos-sotec committed:
fix hono check style

gdimitropoulos-sotec committed:
replace System.exit with Quarkus.asyncExit

gdimitropoulos-sotec committed:
set default quarkus.container-image.push to false

calohmn commented on device-communication/pom.xml:
```diff
@@ -0,0 +1,217 @@
+                    </execution>
+                </executions>
+                <configuration>
+                    <configLocation>checkstyle/default.xml</configLocation>
```
There are still some checkstyle violations being reported when the maven build is run. These refer to generated-sources:
>  target/generated-sources/annotations/org/eclipse/hono/communication/api/mapper/DeviceConfigMapperImpl.java

To ignore the generated-sources for the checkstyle check, the source directories can be defined explicitly here by adding:
```
                    <sourceDirectories>
                        <sourceDirectory>${project.build.sourceDirectory}</sourceDirectory>
                        <sourceDirectory>${project.build.testSourceDirectory}</sourceDirectory>
                    </sourceDirectories>
```

gdimitropoulos-sotec replied:
Done

calohmn commented on device-communication/src/main/java/org/eclipse/hono/communication/api/DeviceCommunicationHttpServer.java:
```diff
@@ -0,0 +1,244 @@
+     * Adds status code 400 and sets the error message for the routingContext response.
+     *
+     * @param routingContext the routing context object
+     * @Throws: NullPointerException – if routingContext is {@code null}.
```
`@throws`, same in the javadoc of the method below.

gdimitropoulos-sotec replied:
Done

calohmn replied:
The colon should also be removed.

gdimitropoulos-sotec replied:
Done

calohmn commented on device-communication/src/main/java/org/eclipse/hono/communication/core/http/HttpServiceBase.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Base HTTP service class.
+ */
+public class HttpServiceBase {
```
This class is not being used. Can it be removed?

gdimitropoulos-sotec replied:
Done

gdimitropoulos-sotec committed:
remove class HttpServiceBase

gdimitropoulos-sotec committed:
remove check style for auto generated mapper impl

gdimitropoulos-sotec committed:
fix typo Throws to throws

calohmn commented:
@gdimitropoulos-sotec Is #21 meant to replace this PR here?

gdimitropoulos-sotec commented:
Hi @calohmn  PR 21 adds the pub-sub functionality and is based on this PR here.

gdimitropoulos-sotec committed:
remove colon from java docs

calohmn commented on device-communication/pom.xml:
```diff
@@ -0,0 +1,221 @@
+    <dependencies>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-arc</artifactId>
```
`quarkus-arc` includes a dependency of `quarkus-ide-launcher`. In Hono, we have excluded the `quarkus-ide-launcher` dependency - it is marked by the eclipse license check tool as not having consistent license information and it didn't seem necessary. Do we need that dependency here? (I tested running the application with Intellij IDEA Ultimate via a Quarkus run configuration - that worked without the dependency.) If not, we could add an exclusion here:
```
            <exclusions>
                <exclusion>
                    <groupId>io.quarkus</groupId>
                    <artifactId>quarkus-ide-launcher</artifactId>
                </exclusion>
            </exclusions>
```
This should then also be added to the `quarkus-vertx` dependency declaration below.

gdimitropoulos-sotec replied:
Done.

gdimitropoulos-sotec committed:
exclude quarkus-ide-launcher

gdimitropoulos-sotec commented:
@calohmn  do you prefer to review this PR and after the 21 or should I create one PR for both?

calohmn commented on device-communication/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,36 @@
+  version: ${COM_APP_VERSION:"v1"}
+vertx:
+  openapi:
+    file: ${COM_OPENAPI_FILE_PATH:/api/hono-device-communication-v1.yaml}
```
It looks like the leading slash needs to be removed (as already done in the Pub/Sub PR).

gdimitropoulos-sotec replied:
Done

calohmn commented on device-communication/src/main/resources/api/hono-endpoint.yaml:
It seems this file isn't being referenced. I guess it would be easier to manage only one API definition file (the one with the openAPI definition).

gdimitropoulos-sotec replied:
This file is the Openapi 3 file converted to the Swagger 2.0 version and is mandatory for the Google Cloud Endpoints.

calohmn commented on device-communication/src/main/java/org/eclipse/hono/communication/api/DeviceCommunicationHttpServer.java:
```diff
@@ -0,0 +1,244 @@
+                    this.startVertxServer(apiRouter);
+                })
+                .onFailure(error -> {
+                    if (error != null) {
+                        log.error(error.getMessage());
+                    } else {
+                        log.error("Can not create Router");
+                    }
```
Just using `log.error("Can not create Router", error);` would make the kind of error and its cause more prominent in the log, I think.

gdimitropoulos-sotec replied:
I changed the if case log to  log.error("Can not create Router {}", error.getMessage());

calohmn commented on device-communication/src/main/java/org/eclipse/hono/communication/api/DeviceCommunicationHttpServer.java:
```diff
@@ -0,0 +1,244 @@
+        apiRouter.errorHandler(400, routingContext ->
+                ResponseUtils.errorResponse(routingContext, routingContext.failure()));
+        apiRouter.errorHandler(404, routingContext ->
+                ResponseUtils.errorResponse(routingContext, routingContext.failure()));
```
`routingContext.failure()` seems to be `null` here, meaning the 404 exception will be mapped to a status 500 response which is rather confusing.
(I got such a 500 error having used a wrong URL path and wondered what the kind of "Internal server error" was, because there is also no log output in this case, which I would have expected for a status 500 error.
A status 404 error wouldn't have made me look in the log output.)

gdimitropoulos-sotec replied:
I just remove these two Handlers.

calohmn commented on device-communication/README.md:
```diff
@@ -0,0 +1,104 @@
+
+#### commands/{tenantId}/{deviceId}
+
+- POST : post a command for a specific device
```
A hint that this isn't implemented yet would be good here.

gdimitropoulos-sotec replied:
Done

calohmn commented on device-communication/mvnw:
This file should have the executable flag set.

gdimitropoulos-sotec replied:
I make the file executable, but I think there is a bug on Windows, and git update-index --chmod=+x is not working.

calohmn replied:
I've made a corresponding change in c01bccb2b2d9c610cbcb0972486c7a8967bb1a57.

calohmn commented:
> do you prefer to review this PR and after the 21 or should I create one PR for both?

Sorry for the delay. I've added a last set of comments. Then we can merge this.
(If you prefer, you could also go ahead with #21, abandoning this PR. On the other hand, it's good to have smaller PRs,
and as I've understood, #21 will need to wait for the upcoming Hono 2.4.0, so we can start with this PR here first.)

gdimitropoulos-sotec committed:
remove leading slash from openapi file

gdimitropoulos-sotec committed:
fix log error

gdimitropoulos-sotec committed:
remove router error handlers

gdimitropoulos-sotec committed:
add hint for post command

gdimitropoulos-sotec committed:
make mvnw executable

gdimitropoulos-sotec commented:
> Sorry for the delay. I've added a last set of comments. Then we can merge this. (If you prefer, you could also go ahead with #21, abandoning this PR. On the other hand, it's good to have smaller PRs, and as I've understood, #21 will need to wait for the upcoming Hono 2.4.0, so we can start with this PR here first.)


Thank you for your time. I have fixed all the issues from your comments and also update the other PR with these changes.

calohmn reviewed (approved):
LGTM, thanks for contributing.

calohmn commented:
@gdimitropoulos-sotec As a follow-up, I've adapted the GitHub Actions workflow to also build the device communication API. I've also made some small adaptations/corrections - see c01bccb2b2d9c610cbcb0972486c7a8967bb1a57. This also includes choosing a generic application docker image name and setting the `quarkus.container-image.build` flag to `false` by default. I think it's better to let the user explicitly choose to build the docker image (like in Hono where an extra maven profile is used) and adapt the docker-registry/organization name along with that.
